http POST based rpc support.

balamurugana opened the pull request:
Added support for new RPC support using HTTP POST.  RPC's argumnets
and reply are Gob encoded and sent as HTTP request/response body.

This patch also removes Go RPC based implementation.

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -245,7 +245,7 @@ func (a adminAPIHandlers) ServerInfoHandler(w http.ResponseWriter, r *http.Reque
 			reply[idx] = ServerInfo{Addr: peer.addr}
 
-			serverInfoData, err := peer.cmdRunner.ServerInfoData()
+			serverInfoData, err := peer.cmdRunner.ServerInfo()
```
@balamurugana it is better than you avoid naming changes, I expect to only see that HTTP based RPC support in this PR. If you wish to do that you can do that as a separate PR.

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -808,21 +808,20 @@ func (a adminAPIHandlers) UpdateCredentialsHandler(w http.ResponseWriter,
 	}
 
+	// Notify all other Minio peers to update credentials
+	for host, err := range globalNotificationSys.SetCredentials(creds) {
```
I see that you converted admin RPC also into notification sub-system, I would also suggest that you make this change in a separate PR.

harshavardhana commented on cmd/posix.go:
```diff
@@ -142,26 +142,23 @@ func isDirEmpty(dirname string) bool {
 
 // Initialize a new storage disk.
-func newPosix(path string) (StorageAPI, error) {
+func newPosix(path string) (*posix, error) {
```
Why do you wish to return `*posix` ?

harshavardhana commented on cmd/rpc-version.go:
```diff
@@ -0,0 +1,41 @@
+package cmd
```
Missing license..

harshavardhana commented on pkg/rpc/client.go:
```diff
@@ -0,0 +1,101 @@
+	return &Client{
+		httpClient: &http.Client{
+			Transport: &http.Transport{
+				IdleConnTimeout: idleTimeout,
```
We need to bring in `MaxIdleConnsPerHost` to higher than 2 to avoid TIME_WAIT build up.

balamurugana replied:
unable to understand. default idle timeout is 5 minutes as per my AWS S3 testing.

harshavardhana replied:
I am not talking about idleTimeout - I am talking about connection pooling behavior.

balamurugana replied:
Unable to understand your question.

harshavardhana replied:
MaxIdleConnsPerHost is one of the options for http.Transport default value to '2' leads to a large number of connections in TIME_WAIT when we have a high concurrent workload. It is better to move that value to a higher number. In my past tests `MaxIdleConnsPerHost: 100` has turned out to be a good value.  Some similar findings by others http://tleyden.github.io/blog/2016/11/21/tuning-the-go-http-client-library-for-load-testing/

I will re-test this anyways, to recommend default values here.

harshavardhana commented on pkg/rpc/client.go:
```diff
@@ -0,0 +1,101 @@
+	return nil
+}
+
+// NewClient - returns new RPC client.
```
should we not say http based RPC client?

balamurugana replied:
It is minio specific rpc client (not general purpose any more) which is anyway http post. I don't think its required to specify

harshavardhana commented on pkg/rpc/client.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+// NewClient - returns new RPC client.
+func NewClient(serviceURL *xnet.URL, tlsConfig *tls.Config, idleTimeout time.Duration) *Client {
```
you are better off takin http.Client directly, we would need to provide more configurable option anyways in the near future. It is better to establish that caller has full control over their http.Client - they can even provide custom traceable http.Transport etc in this manner.

balamurugana replied:
If any such requirement is in future, that will be done when needed.

harshavardhana replied:
if you want to make this a generic package and not break compatibility you should think about that. The current requirement is not going to address the concerns I have about connection pooling. I will test this and recommend values as we agreed.

harshavardhana commented on cmd/storage-rpc-client.go:
```diff
@@ -18,34 +18,25 @@ package cmd
+func isNetworkDisconnectError(err error) bool {
+	if err != nil {
+		_, isNetOpError := err.(*net.OpError)
+		return isNetOpError || err == rpc.ErrShutdown
```
There is no rpc.ErrShutdown anymore I believe.

harshavardhana commented on cmd/lock-rpc-client.go:
```diff
@@ -16,56 +16,99 @@
+
+	rpcClient, err := NewRPCClient(
+		RPCClientArgs{
+			NewAuthTokenFunc: newAuthToken,
```
I am not able to see the authToken being set in any headers, as a client we should set it as part of Authorization header such that server can validate. In this change, there is no authentication as such.  please correct me if i am wrong.

balamurugana replied:
http POST style brings different way of same authentication.  I would like to setup a call to discuss in detail.

harshavardhana reviewed (changes requested):
@balamurugana there are a lot of unrelated changes in this PR, I am not sure why. Here is my list of suggestions

- Changing names for rpc methods (can you send a separate PR for that)
- Moving existing admin-rpc, s3-peer rpc into globalNotificationSys methods (send a separate PR)
- Simplification of rpc registrations to use xnet package (send a separate PR)
- Then rebase your changes for HTTP based RPC on top of that. 

If you send separate PRs then it is easier to review, otherwise the whole change to move from net/rpc to net/http doesn't justify all the cleanups that were done along with it.

balamurugana committed:
remove old rpc code

balamurugana committed:
add new HTTP POST based rpc code

balamurugana committed:
changes in admin rpc server/client code

balamurugana committed:
changes in lock rpc server/client code

balamurugana committed:
changes in peer rpc server/client code

balamurugana committed:
changes in storage rpc server/client code

balamurugana committed:
HTTP POST based RPC support.

Added support for new RPC support using HTTP POST.  RPC's argumnets
and reply are Gob encoded and sent as HTTP request/response body.

This patch also removes Go RPC based implementation.

changes in other global codes.

krishnasrinivas commented on cmd/admin-rpc-server.go:
```diff
@@ -18,223 +18,110 @@ package cmd
+var adminServicePath = path.Join(minioReservedBucketPath, peerServiceSubPath)
+
+// adminRPCReceiver - Admin RPC receiver for admin RPC server.
+type adminRPCReceiver struct {
```
adminRPCServer is a better name?

balamurugana replied:
Based on Go RPC terms, I used receiver than server.  At server side, two components required at server side.  Receiver type contains methods for RPCs and RPC server receives/serves RPC request/response.  The server contains one or more receivers (not confirmed/tested with 2+ receivers) to serve RPCs. Hence both are two different thing.

krishnasrinivas replied:
Got it. But...

`AdminRPCClient` and `AdminRPCServer` looks complementary and predictable names.

`AdminRPCClient` and `AdminRPCReceiver` does not sound complementary.

Since we are not using Go RPC anymore, we could use the names that are more meaningful for us.

> The server contains one or more receivers (not confirmed/tested with 2+ receivers) to serve RPCs. Hence both are two different thing.

For the RPC package, it can think of `adminRPCServer` as a sub-rpc-server instead of an unfamiliar term like `receiver`

krishnasrinivas commented on cmd/admin-rpc-server.go:
```diff
@@ -18,223 +18,110 @@ package cmd
-// restart commands.
-type adminCmd struct {
-	AuthRPCServer
+var adminServicePath = path.Join(minioReservedBucketPath, peerServiceSubPath)
```
admin commands are failing, this should be 
```
var adminServicePath = path.Join(minioReservedBucketPath, adminServiceSubPath)
```

balamurugana replied:
Done

krishnasrinivas reviewed (commented):
We use the key words `client` `server` `service` `receiver`, instead we could just use `client` and `server` so that it is simple and predictable.
